src/zarr/api/asynchronous.py

```diff
@@ -225,6 +225,21 @@ async def open(
     if path is not None:
         store_path = store_path / path
 
+    if "shape" not in kwargs and mode in {"a", "w", "w-"}:
+        try:
+            metadata_dict = await get_array_metadata(store_path, zarr_format=zarr_format)
+            zarr_format = metadata_dict["zarr_format"]
+            if zarr_format == 3:
+                is_array = metadata_dict.get("node_type") == "array"
+            else:
+                # for v2, the above would already have raised an exception if not an array
+                is_array = True
+        except (AssertionError, FileNotFoundError):
+            is_array = False
+        if not is_array:
+            return await open_group(store=store_path, zarr_format=zarr_format, mode=mode, **kwargs)
+        else:
+            return await open_array(store=store_path, zarr_format=zarr_format, mode=mode, **kwargs)
```

Comment:
it seems like we should be able to use the `metadata_dict` to construct the array without calling `open_array` and repeating the IO step, no?